8319196: ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode

cushon opened the pull request:
Hello,

Please consider this fix for [JDK-8319196: ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode](https://bugs.openjdk.org/browse/JDK-8319196).

As discussed in [the accompanying CSR](https://bugs.openjdk.org/browse/JDK-8319330), `ExecutableElement.getReceiverType()` and `ExecutableType.getReceiverType()` currently returns `NoType` unless the receiver parameter syntax was present in source, which also means that `getReceiverType()` doesn't work for methods loaded from class files.

This change updates the implementation to check if type annotations are present on the receiver type of a method read from bytecode, and fill in the receiver type of that method to host the type annotations.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330) to be approved

### Issues
 * [JDK-8319196](https://bugs.openjdk.org/browse/JDK-8319196): ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode (**Bug** - P4)
 * [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330): ExecutableElement.getReceiverType doesn't return receiver types for methods loaded from bytecode (**CSR**)


### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [32f0050a](https://git.openjdk.org/jdk/pull/16482/files/32f0050a70fe2b807090d5698dfd61339c503880)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16482/head:pull/16482` \
`$ git checkout pull/16482`

Update a local copy of the PR: \
`$ git checkout pull/16482` \
`$ git pull https://git.openjdk.org/jdk.git pull/16482/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16482`

View PR using the GUI difftool: \
`$ git pr show -t 16482`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16482.diff">https://git.openjdk.org/jdk/pull/16482.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16482#issuecomment-1791281930)

liach commented on src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:
```diff
@@ -2276,6 +2277,21 @@ public Type getReceiverType() {
+            if (enclosingClass == null) {
+                return null;
+            }
+            Type enclosingType = enclClass().type;
```
```suggestion
            Type enclosingType = enclosingClass.type;
```

liach commented on src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:
```diff
@@ -2276,6 +2277,21 @@ public Type getReceiverType() {
+            if (isConstructor()) {
+                return enclosingType.getEnclosingType();
+            }
+            if (!isStatic()) {
```
Just curious, what happens in such a scenario:
```java
public class Outer {
  public static void work() {
    class Local {
      Local() {} // <- what's the implicitReceiverType here?
    }
  }
}
```
FYI javac currently is not marking such local classes in static methods as `static` in its `InnerClasses` attribute.

cushon replied:
Good question! I debugged that example, and `getEnclosingType()` returns `NoType`, which I think is what we want.

This is the logic that initializes the enclosing type and handles local classes:

https://github.com/openjdk/jdk/blob/c788160f8acea7b58b54ad857b601bb7ffb53f8e/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Enter.java#L508-L519

cushon commented:
I have tentatively updated this to only fill in a receiver type during class reading if type annotations are present on the receiver, based on [discussion in the CSR](https://bugs.openjdk.org/browse/JDK-8319330?focusedId=14624193&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14624193).

vicente-romero-oracle commented on test/langtools/tools/javac/processing/model/element/TestExecutableReceiverType.java:
```diff
@@ -45,8 +46,14 @@ public boolean process(Set<? extends TypeElement> annotations,
-            count += testType(elements.getTypeElement("MethodHost.Nested"));
+            for (ExecutableElement e : ElementFilter.methodsIn(
+                  roundEnv.getElementsAnnotatedWith(ReceiverTypeKind.class))) {
+              count += testExecutable(e);
```
nit: indentation should be 4 spaces here, same issue below

cushon replied:
Done

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/code/Symbol.java:
```diff
@@ -81,6 +81,7 @@
 import static com.sun.tools.javac.jvm.ByteCodes.ishll;
 import static com.sun.tools.javac.jvm.ByteCodes.lushrl;
 import static com.sun.tools.javac.jvm.ByteCodes.lxor;
+import static com.sun.tools.javac.jvm.ByteCodes.ret;
```
this import doesn't seem to be used now

cushon replied:
Done

vicente-romero-oracle reviewed (approved):
looks good to me, just minor comments

cushon commented:
> looks good to me, just minor comments

I addressed the comments, thanks for the review!

I also updated the CSR [JDK-8319330](https://bugs.openjdk.org/browse/JDK-8319330) and moved it back to 'proposed'.

cushon commented:
/integrate
